X3: rollback iterator when rule's handler returns fail

Xeverous opened the pull request:
Test attempt to fix #833. First checking if I didn't broke anything

Xeverous committed:
X3: rollback iterator when rule fails to parse

This commit adjusts work of commit 2db3fde.
It corrected error_handler which should not skip whitespaces but
introduced a regression where handlers returning fail were still
moving the iterator forward. This commit adds iterator rollback
and synchronizes rule's and guard's iterators behavior.

saki7 commented:
You can ignore the failures on AppVeyor. Our Windows CI is not properly configured since May 2025 and it has nothing to do with your PR. I'm not planning to fix this soon because all my CI renewals are reverted due to the C++23 revert.

Xeverous commented:
Any opinion from you on the changes in code? The only implementation idea that came to my mind is to simply rollback the iterator when `fail` is returned from the handler (basically, turn `>` into `>>`).

I noticed there are comments in `rule.hpp` and `guard.hpp` to keep them in sync. Even before my change, the code differed in regards to how it treats iterators. The guard implementation makes a copy of the iterator before parsing call (call with the copy) and only sets proper iterator if the parse succeeded.

The rule implementation did not have this iterator copy and I wasn't sure whether:
A) it should rollback only on fail (this is what I did in this patch)
B) always rollback, except when parse succeeds (this is what guard implementation does)

saki7 commented:
You should sync the implementations of rule and guard. I think B is safer. You can consider this as the exception for "don't touch anything except for rule" I mentioned.

Xeverous commented:
Adjusted the commit to solution B and added a description into the commit message.

Unfortunately X3 has no `error_handler_result` tests that could be used to infer any particular design of iterator behavior.

My intuition:
- `error_handler_result::fail`: rollback iterator
- `error_handler_result::retry`: (?) keep modified one (to retry in a different place), otherwise (unless the user changes some state) it will be applied forever in a loop; currently it rolls back
- `error_handler_result::accept`: do not rollback - this is currently not implemented
- `error_handler_result::rethrow`: with exceptions it does rethrow, without there is a TODO and I'm not really sure what should happen

saki7 commented:
The commit message looks very good, it addresses the investigation requirement I mentioned in https://github.com/boostorg/spirit/issues/833#issuecomment-3337339192.

I'd rather choose to not touch the irrelevant enumerations other than `::fail`, thanks for the fix anyway.
